File: frontend/src/utils/provenance.js
/**
 * provenance.js — input provenance metadata helpers.
 *
 * Implements the v2.1 state contract's provenance schema. Per the contract:
 *   - Every input in {building, gains, operation, systems}_config has an
 *     associated provenance record tracking where the value came from.
 *   - Records live in a sibling `_provenance` block keyed by dot-notated
 *     input path (e.g. "fabric.external_wall.u_value").
 *   - Per-path record: { source, ref?, confidence?, recorded_at? }
 *   - `source` is required and one of the six contract enum values.
 *   - Default when unspecified: { source: 'user_entered', confidence: 'medium' }.
 *
 * States 1–3 record provenance but don't branch on it; State 4 reconciliation
 * reads it to weight the bottom-up estimate and bound proposed adjustments.
 *
 * See `docs/state_contracts.md` § Cross-cutting concepts → Input provenance.
 */

/** Canonical provenance source enum, per contract v2.1. */
export const PROVENANCE_SOURCES = Object.freeze([
  'user_entered',
  'spec_sheet',
  'vintage_default',
  'benchmark',
  'inferred',
  'calibrated',
])

/** Default record returned by getProvenance when no entry exists at the path. */
export const DEFAULT_PROVENANCE = Object.freeze({
  source: 'user_entered',
  confidence: 'medium',
})

/**
 * Get the provenance record for a given input path.
 *
 * @param {object|null|undefined} config — a config blob (building_config, etc.).
 * @param {string} path — dot-notated path of the input within the config.
 * @returns {{source, ref?, confidence?, recorded_at?}} — the stored record,
 *          or the contract default if none is set / config is null.
 */
export function getProvenance(config, path) {
  if (!config || !path) return { ...DEFAULT_PROVENANCE }
  const prov = config._provenance
  if (!prov || typeof prov !== 'object') return { ...DEFAULT_PROVENANCE }
  const rec = prov[path]
  if (!rec || typeof rec !== 'object' || !rec.source) {
    return { ...DEFAULT_PROVENANCE }
  }
  return { ...rec }
}

/**
 * Set the provenance record for a given input path. Immutable — returns a
 * new config object with the `_provenance` block updated.
 *
 * @param {object} config — a config blob (must be non-null; pass {} for empty).
 * @param {string} path — dot-notated path of the input.
 * @param {{source, ref?, confidence?, recorded_at?}} record — must include
 *        `source` (one of PROVENANCE_SOURCES). Throws otherwise.
 * @returns {object} new config with the record merged into _provenance.
 */
export function setProvenance(config, path, record) {
  if (!config || typeof config !== 'object') {
    throw new Error('setProvenance: config must be an object')
  }
  if (!path || typeof path !== 'string') {
    throw new Error('setProvenance: path must be a non-empty string')
  }
  if (!record || !record.source) {
    throw new Error('setProvenance: record.source is required')
  }
  if (!PROVENANCE_SOURCES.includes(record.source)) {
    throw new Error(
      `setProvenance: source '${record.source}' is not a valid contract enum value ` +
      `(must be one of: ${PROVENANCE_SOURCES.join(', ')})`
    )
  }
  const cleaned = {
    source: record.source,
    ...(record.ref !== undefined ? { ref: record.ref } : {}),
    ...(record.confidence !== undefined ? { confidence: record.confidence } : {}),
    recorded_at: record.recorded_at ?? new Date().toISOString(),
  }
  return {
    ...config,
    _provenance: {
      ...(config._provenance ?? {}),
      [path]: cleaned,
    },
  }
}

/**
 * Remove a single provenance entry (so subsequent reads fall back to default).
 * Immutable.
 *
 * @param {object} config
 * @param {string} path
 * @returns {object} new config with the entry removed.
 */
export function clearProvenance(config, path) {
  if (!config || typeof config !== 'object') return config
  if (!path || !config._provenance) return config
  if (!(path in config._provenance)) return config
  const nextProv = { ...config._provenance }
  delete nextProv[path]
  // If the _provenance block is now empty, drop it entirely.
  if (Object.keys(nextProv).length === 0) {
    const { _provenance: _drop, ...rest } = config
    return rest
  }
  return { ...config, _provenance: nextProv }
}

/**
 * Enumerate every provenance entry currently set on a config.
 *
 * @param {object|null|undefined} config
 * @returns {Array<{path, source, ref?, confidence?, recorded_at?}>}
 */
export function listProvenance(config) {
  if (!config || !config._provenance || typeof config._provenance !== 'object') {
    return []
  }
  return Object.entries(config._provenance)
    .filter(([_, rec]) => rec && rec.source)
    .map(([path, rec]) => ({ path, ...rec }))
}

File: nza_engine/utils/provenance.py
"""
provenance.py — input provenance metadata helpers.

Mirror of `frontend/src/utils/provenance.js`. Implements the v2.1 state
contract's provenance schema; the two implementations must stay in lockstep
because both backend and frontend read the same `building_config` /
`gains_config` / etc. JSON blobs.

Per the contract (`docs/state_contracts.md` § Cross-cutting concepts →
Input provenance):
    - Provenance lives in a sibling `_provenance` object keyed by
      dot-notated input path (e.g. "fabric.external_wall.u_value").
    - Per-path record: { source, ref?, confidence?, recorded_at? }
    - `source` is required, one of the six contract enum values.
    - Default when unspecified: { source: 'user_entered', confidence: 'medium' }.

States 1–3 record provenance but don't branch on it. State 4 reconciliation
reads it to weight the bottom-up estimate and bound proposed adjustments.
"""

from __future__ import annotations

from copy import deepcopy
from datetime import datetime, timezone
from typing import Any


# Canonical provenance source enum, per contract v2.1.
PROVENANCE_SOURCES: tuple[str, ...] = (
    "user_entered",
    "spec_sheet",
    "vintage_default",
    "benchmark",
    "inferred",
    "calibrated",
)

# Default record returned by get_provenance when no entry exists at the path.
DEFAULT_PROVENANCE: dict[str, Any] = {
    "source": "user_entered",
    "confidence": "medium",
}


def get_provenance(config: dict | None, path: str) -> dict[str, Any]:
    """Read the provenance record for a given input path.

    Falls back to DEFAULT_PROVENANCE if no entry is set or config is None.
    Returns a fresh dict so callers can't mutate the default.
    """
    if not config or not path:
        return dict(DEFAULT_PROVENANCE)
    prov = config.get("_provenance")
    if not isinstance(prov, dict):
        return dict(DEFAULT_PROVENANCE)
    rec = prov.get(path)
    if not isinstance(rec, dict) or not rec.get("source"):
        return dict(DEFAULT_PROVENANCE)
    return dict(rec)


def set_provenance(config: dict, path: str, record: dict[str, Any]) -> dict:
    """Write or replace a provenance record. Immutable — returns a new config.

    `record.source` is required and must be one of PROVENANCE_SOURCES. Other
    fields (ref, confidence, recorded_at) are optional. If recorded_at is
    absent it's filled with the current UTC ISO 8601 timestamp.
    """
    if not isinstance(config, dict):
        raise ValueError("set_provenance: config must be a dict")
    if not path or not isinstance(path, str):
        raise ValueError("set_provenance: path must be a non-empty string")
    if not isinstance(record, dict) or not record.get("source"):
        raise ValueError("set_provenance: record.source is required")
    if record["source"] not in PROVENANCE_SOURCES:
        raise ValueError(
            f"set_provenance: source {record['source']!r} is not a valid contract enum "
            f"value (must be one of: {', '.join(PROVENANCE_SOURCES)})"
        )

    cleaned: dict[str, Any] = {"source": record["source"]}
    if "ref" in record and record["ref"] is not None:
        cleaned["ref"] = record["ref"]
    if "confidence" in record and record["confidence"] is not None:
        cleaned["confidence"] = record["confidence"]
    cleaned["recorded_at"] = record.get("recorded_at") or datetime.now(timezone.utc).isoformat()

    next_config = deepcopy(config)
    prov = next_config.get("_provenance")
    if not isinstance(prov, dict):
        prov = {}
    prov[path] = cleaned
    next_config["_provenance"] = prov
    return next_config


def clear_provenance(config: dict, path: str) -> dict:
    """Remove a single entry so subsequent reads return the default. Immutable."""
    if not isinstance(config, dict) or not path:
        return config
    prov = config.get("_provenance")
    if not isinstance(prov, dict) or path not in prov:
        return config
    next_config = deepcopy(config)
    next_prov = dict(next_config["_provenance"])
    del next_prov[path]
    if next_prov:
        next_config["_provenance"] = next_prov
    else:
        del next_config["_provenance"]
    return next_config


def list_provenance(config: dict | None) -> list[dict[str, Any]]:
    """Enumerate every provenance entry currently set on a config.

    Returns a list of { path, source, ref?, confidence?, recorded_at? } dicts.
    Entries missing the required `source` field are skipped silently.
    """
    if not isinstance(config, dict):
        return []
    prov = config.get("_provenance")
    if not isinstance(prov, dict):
        return []
    out: list[dict[str, Any]] = []
    for path, rec in prov.items():
        if not isinstance(rec, dict) or not rec.get("source"):
            continue
        out.append({"path": path, **rec})
    return out
